libs/vkd3d/swapchain.c

```diff
@@ -0,0 +1,2216 @@
+        VkPhysicalDevice vk_physical_device, uint32_t memory_type_mask,
+        VkMemoryPropertyFlags flags, uint32_t *memory_type_index)
+{
+    VkPhysicalDeviceMemoryProperties memory_properties;
+    unsigned int i;
+
+    vk_procs->vkGetPhysicalDeviceMemoryProperties(vk_physical_device, &memory_properties);
+    for (i = 0; i < memory_properties.memoryTypeCount; ++i)
+    {
+        if (!(memory_type_mask & (1u << i)))
+            continue;
+
+        if ((memory_properties.memoryTypes[i].propertyFlags & flags) == flags)
+        {
+            *memory_type_index = i;
+            return S_OK;
+        }
+    }
+
+    FIXME("Failed to find memory type (allowed types %#x).\n", memory_type_mask);
+    return E_FAIL;
+}
+
+static BOOL d3d12_swapchain_is_present_mode_supported(struct d3d12_swapchain *swapchain,
+        VkPresentModeKHR present_mode)
+{
+    VkPhysicalDevice vk_physical_device = d3d12_swapchain_device(swapchain)->vk_physical_device;
+    const struct vkd3d_vk_device_procs *vk_procs = d3d12_swapchain_procs(swapchain);
+    VkPresentModeKHR *modes;
+    uint32_t count, i;
+    BOOL supported;
+    VkResult vr;
+
+    if (present_mode == VK_PRESENT_MODE_FIFO_KHR)
+        return TRUE;
+
+    if ((vr = vk_procs->vkGetPhysicalDeviceSurfacePresentModesKHR(vk_physical_device,
+            swapchain->vk_surface, &count, NULL)) < 0)
+    {
+        WARN("Failed to get count of available present modes, vr %d.\n", vr);
+        return FALSE;
+    }
+
+    supported = FALSE;
+
+    if (!(modes = vkd3d_calloc(count, sizeof(*modes))))
+        return FALSE;
+    if ((vr = vk_procs->vkGetPhysicalDeviceSurfacePresentModesKHR(vk_physical_device,
+            swapchain->vk_surface, &count, modes)) >= 0)
+    {
+        for (i = 0; i < count; ++i)
+        {
+            if (modes[i] == present_mode)
+            {
+                supported = TRUE;
+                break;
+            }
+        }
+    }
+    else
+    {
+        WARN("Failed to get available present modes, vr %d.\n", vr);
+    }
+    vkd3d_free(modes);
+
+    return supported;
+}
+
+static BOOL d3d12_swapchain_has_user_images(struct d3d12_swapchain *swapchain)
+{
+    return !!swapchain->vk_images[0];
+}
+
+static HRESULT d3d12_swapchain_create_user_buffers(struct d3d12_swapchain *swapchain, VkFormat vk_format)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = d3d12_swapchain_procs(swapchain);
+    VkPhysicalDevice vk_physical_device = d3d12_swapchain_device(swapchain)->vk_physical_device;
+    VkDevice vk_device = d3d12_swapchain_device(swapchain)->vk_device;
+    VkImageCreateInfo image_info;
+    unsigned int i;
+    VkResult vr;
+    HRESULT hr;
+
+    if (d3d12_swapchain_has_user_images(swapchain))
+        return S_OK;
+
+    memset(&image_info, 0, sizeof(image_info));
+    image_info.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
+    image_info.flags = VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT;
```

Comment:
This breaks DCC unless we use image_format_list, did Wine do this, is there a requirement that we can create arbitrarily aliased views in D3D12?

Reply:
Wine did indeed do this, and we need to be able to cast formats within the same format family.

It's probably better to create these images using existing vkd3d functions (`vkd3d_image_resource_create_info`) since format compatibility will be handled by that.

Reply:
Changed to use vkd3d resource creation.